### PR TITLE
roachpb: introduce the concept of `SystemSpanConfig` and related protos

### PR DESCRIPTION
**Pull request by arulajmani**
This patch is motivated by the desire to let the host tenant lay
protected timestamps on one or all secondary tenants' keyspace. It
also provides a mechanism to allow secondary tenants to lay protected
timestamps on their entire keyspace without updating every span
configuration.

We introduce the concept of `SystemSpanConfig` and
`SystemSpanConfigTarget` to enable this. We tie these together using a
`SystemSpanConfigEntry`.

A `SystemSpanConfig` is a system installed configuration that can apply
to multiple spans. It only contains protected timestamp information.

A `SystemSpanConfigTarget` is used to specify the spans a
`SystemSpanConfig` applies over. It can be used to target the entire
(logical) cluster or a particular secondary tenant. We will ensure that
only the host tenant can target secondary tenants in a future PR that
actually persists `SystemSpanConfigs`.

We will persist `SystemSpanConfigs` in `system.span_configurations` in
a future patch. The `SystemSpanConfigTarget` will be encoded into
special reserved keys when we do so.

This change introduces the notion of a hierarchy to span configurations.
The configuration that applies to a span will now bee the `SpanConfig`
stored in `system.span_configurations` combined with all the
`SystemSpanConfigs` that apply to the span. This can be at most 4
levels deep -- for a secondary tenant's range, the secondary tenant can
install a `SystemSpanConfig` that applies to all its ranges, the host
tenant can install a `SystemSpanConfig` that applies to all ranges of
the secondary tenant, and the host tenant can install a
`SystemSpanConfig` that applies to all ranges.

These protos form the data model which will later be used to enable
protected timestamp support for secondary tenants using the span config
infrastructure. It will be used by the various components such as the
`SQLTranslator`, `KVAccessor`, `Reconciler` etc.

Release note: None

### CODE REVIEW

**Comment by arulajmani** [comment]
This is slightly different from the data model described in the [RFC](https://github.com/cockroachdb/cockroach/pull/74685), prompted by a discussion I had with @irfansharif and @adityamaru offline after the RFC was published. Specifically the introduction of the `SystemSpanConfigTarget` message -- by doing so, we can keep the meaning of the special keys we are hoping to reserve contained within `KVAccessor` and `KVSubscriber`. The `SQLTranslator` and `Reconciler` won't need to know how `SystemSpanConfigs` are stored internally. 

I'll go back and update the RFC with where this PR ends up.

File: pkg/roachpb/span_config.proto
Comment (adityamaru) [inline pkg/roachpb/span_config.proto:41]: I think this field should be ID 1?

**Review by adityamaru — APPROVED** [review]
This matches my mental model, but again I might be too close to the fire so we should probably wait for one more LGTM.

**Comment by arulajmani** [comment]
@nvanbenschoten could I get a look on this from you as well before we merge this? Specifically the discussion around the introduction of the `SystemSpanConfig` message that you raised on the RFC. Is my rationale (copy pasted from the RFC discussion below) convincing? 

> We considered using the same SpanConfig proto but decided to introduce this special type once the design moved away from generalized inheritance. As the hierarchy we're proposing is strictly above RANGE DEFAULT, SpanConfigs are guaranteed to be fully hydrated so we can never have inheritance of fields. We made this fully hydrated distinction explicit by choosing to make all fields on SpanConfigs non-nullable (unlike zone configs) and introducing this new message lets us keep things that way.
>
> Now that I type this, maybe this fully hydrated argument may not apply in the future if we introduce new fields on SpanConfigs to be only set by the host tenant and apply to all tenant ranges. But a different type here shouldn't preclude us from doing so in the future.

File: pkg/roachpb/span_config.proto
Comment (irfansharif) [inline pkg/roachpb/span_config.proto:182]: Discussed a bit offline, but could this perhaps be pared down to something like the following?

```proto
message SystemSpanConfigTarget {
   // TenantID indicates the tenant ID of the logical cluster being targeted.
   // For secondary tenants this field is left unset. For the host we can use
   // this field to protect a specific guest tenant. Down in KV we already have
   // access to roachpb.TenantFromContext to disambiguate where the request is
   // originating from -- precisely to avoid needing to add data to our message
   // types to encode as much. We can auth that this field is left empty if
   // originating from a secondary tenant.
   roachpb.TenantID tenant_id = 1 [(gogoproto.customname) = "TenantID", (gogoproto.nullable) = true];
 }
```

A relevant example (not too different from what spanconfig.KVAccessor will need to do):

https://github.com/cockroachdb/cockroach/blob/2569e853358070c83e637de214e4d85a834cacdb/pkg/kv/kvserver/replica_rate_limit.go#L27-L28

File: pkg/roachpb/span_config.proto
Comment (irfansharif) [inline pkg/roachpb/span_config.proto:159]: Instead of making the ProtectionPolicy itself a repeated type, I think we could nest the repetition within the message itself and be just as future proof (we only care about repeated timestamps for now, right?). I'd actually go as far as to say perhaps this should be nested with our existing GCPolicy type; these timestamps are only ever used in the context of GC after all.

**Review by irfansharif — COMMENTED** [review]
I'll defer to y'all about whether we want a dedicated SystemSpanConfig type, or just re-use the SpanConfig one.

**Review by arulajmani — COMMENTED** [review]
I spoke to @nvanbenschoten about dedicated `SystemSpanConfig` type vs. not and we agreed that we're on the same page with introducing the new one. 

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/74765#-:-Mty67jfA0vu2q3vK6tg:b-okkiq4)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @adityamaru, @ajwerner, @irfansharif, and @nvanbenschoten)

___
*[pkg/roachpb/span_config.proto, line 41 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74765#-MtFBLQVMg2ct5hEJ0OZ-r2-41:-Mty673Q5sSrNOr9cSPq:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1939e9a74a2648ea9a9cc06376eb964163b3ad5/pkg/roachpb/span_config.proto#L41)):*
<details><summary><i>Previously, adityamaru (Aditya Maru) wrote…</i></summary><blockquote>

I think this field should be ID 1?
</blockquote></details>

Done.
___
*[pkg/roachpb/span_config.proto, line 159 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74765#-MtFBLQVMg2ct5hEJ0OZ-r2-159:-Mty5RDO9ojYUUQuo2qM:bys3f7u) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1939e9a74a2648ea9a9cc06376eb964163b3ad5/pkg/roachpb/span_config.proto#L159)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Instead of making the ProtectionPolicy itself a repeated type, I think we could nest the repetition within the message itself and be just as future proof (we only care about repeated timestamps for now, right?). I'd actually go as far as to say perhaps this should be nested with our existing GCPolicy type; these timestamps are only ever used in the context of GC after all.
</blockquote></details>

We talked about why we want to make `ProtectionPolicy` a repeated type offline -- imagine wanting to introduce `ProtectionMode` in the future for realsies, in which case, we would want to tie the mode to a particular timestamp (as opposed to all timestamps) . The desire here is to future proof ourselves for changes like these. 

That being said, it makes sense to nest this under the `GCPolicy` message because it's only ever used in conjunction with it; done. 
___
*[pkg/roachpb/span_config.proto, line 182 at r2](https://reviewable.io/reviews/cockroachdb/cockroach/74765#-MtFBLQVMg2ct5hEJ0OZ-r2-182:-Mty61LTBrl9VA33vrBE:b-rxji9i) ([raw file](https://github.com/cockroachdb/cockroach/blob/d1939e9a74a2648ea9a9cc06376eb964163b3ad5/pkg/roachpb/span_config.proto#L182)):*
<details><summary><i>Previously, irfansharif (irfan sharif) wrote…</i></summary><blockquote>

Discussed a bit offline, but could this perhaps be pared down to something like the following?

```proto
message SystemSpanConfigTarget {
   // TenantID indicates the tenant ID of the logical cluster being targeted.
   // For secondary tenants this field is left unset. For the host we can use
   // this field to protect a specific guest tenant. Down in KV we already have
   // access to roachpb.TenantFromContext to disambiguate where the request is
   // originating from -- precisely to avoid needing to add data to our message
   // types to encode as much. We can auth that this field is left empty if
   // originating from a secondary tenant.
   roachpb.TenantID tenant_id = 1 [(gogoproto.customname) = "TenantID", (gogoproto.nullable) = true];
 }
```

A relevant example (not too different from what spanconfig.KVAccessor will need to do):

https://github.com/cockroachdb/cockroach/blob/2569e853358070c83e637de214e4d85a834cacdb/pkg/kv/kvserver/replica_rate_limit.go#L27-L28
</blockquote></details>

Thanks for pointing me to this `TenantFromContext` pattern -- done!

File: pkg/roachpb/span_config.proto
Comment (irfansharif) [inline pkg/roachpb/span_config.proto:33]: [nit] Maybe another sentence/phrase about how the data the config applies to is not GC-ed provided it wasn't already GC-ed pre config-application?

File: pkg/roachpb/span_config.proto
Comment (irfansharif) [inline pkg/roachpb/span_config.proto:183]: [minor] Maybe a TODO to later assert that this is the case when wiring things up for real.

**Review by irfansharif — APPROVED** [review]
LGTM. @nvanbenschoten and @ajwerner, this PR introduces the base protos to power the rest of #74685, and it'd be worth taking a look to see what you think.

File: pkg/roachpb/span_config.proto
Comment (arulajmani) [reply pkg/roachpb/span_config.proto:184]: Done.

File: pkg/roachpb/span_config.proto
Comment (arulajmani) [reply pkg/roachpb/span_config.proto]: Done.

**Comment by arulajmani** [comment]
Going to merge this for now as we have a few dependent pieces on this, but if there's any comments on this later I'm happy to retroactively address them. 

bors r+

**Pull request merged**
